src/core/eval.ml

```diff
@@ -77,38 +79,55 @@ let snf : (term -> term) -> (term -> term) = fun whnf ->
     | TRef(_)     -> assert false
   in snf
 
-(** Configuration of the reduction engine. *)
-type config =
-  { context : ctxt (** Context of the reduction used for generating metas. *)
-  ; defmap : term VarMap.t (** Variable definitions. *)
-  ; rewrite : bool (** Use user-defined rewrite rules. *)
-  ; problem : problem (** Generated metavariables. *) }
-
-(*let pp_defmap = D.map VarMap.iter pp_var " ≔ " pp_term "; "*)
+type rw_tag = [ `NoBeta | `NoRw | `NoExpand ]
 
-let cfg_of_ctx : ctxt -> bool -> config = fun context rewrite ->
-  {context; defmap = Ctxt.to_map context; rewrite; problem = new_problem()}
-
-let unfold_cfg : config -> term -> term = fun c a ->
-  let a = unfold a in
-  match a with
-  | Vari x ->
-    begin match VarMap.find_opt x c.defmap with
-      | None -> a
-      | Some v -> unfold v
-    end
-  | _ -> a
+(** Configuration of the reduction engine. *)
+module Config = struct
+
+  type t =
+    { context : ctxt
+    (** Context of the reduction used for generating metas. *)
+    ; defmap : term VarMap.t (** Variable definitions. *)
+    ; rewrite : bool (** Whether to apply user-defined rewriting rules. *)
+    ; expand_defs : bool (** Whether to expand definitions. *)
+    ; beta : bool (** Whether to beta-normalise *)
+    ; problem : problem (** Generated metavariables. *) }
+
+  (** [make ?problem ?rewrite c] creates a new configuration with problem
+      [?problem] (being new if not provided), tags [?rewrite] (being empty if
+      not provided) and context [c]. By default, beta reduction and rewriting
+      is enabled for all symbols. *)
+  let make : ?problem:problem -> ?rewrite:rw_tag list -> ctxt -> t =
+  fun ?(problem=new_problem ()) ?(rewrite=[]) context ->
+    let beta = not @@ List.mem `NoBeta rewrite in
+    let expand_defs = not @@ List.mem `NoExpand rewrite in
+    let rewrite = not @@ List.mem `NoRw rewrite in
+    {context; defmap = Ctxt.to_map context; rewrite; expand_defs; beta; problem}
+
+  let unfold : t -> term -> term = fun c a ->
+    let a = unfold a in
+    match a with
+    | Vari x ->
+      begin match VarMap.find_opt x c.defmap with
+        | None -> a
+        | Some v -> unfold v
```

Comment:
Shouldn't this function be recursive? `let rec unfold c a = let a = Term.unfold a in ...`

Reply:
Yes I think it should indeed.